main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala

```diff
@@ -473,6 +474,129 @@ object EntryPoints {
+      *   - Has a valid signature (without type variables, etc.)
+      *
+      */
+    def run(root: TypedAst.Root)(implicit flix: Flix): (TypedAst.Root, List[EntryPointError]) = {
+      val defaultHandlers = root.defs.filter {
+        case (_, defn) => defn.spec.ann.isDefaultHandler
+      }
+      val (validHandlers, errors) = defaultHandlers.map {
+        case (sym, defn) => checkHandler(sym, defn, root)
+      }.partitionMap  {
+        case Validation.Success(validHandler) => Left(validHandler)
+        case Validation.Failure(err) => Right(err)
+      }
+      val duplicateHandlersErrs = validHandlers.groupBy(_.handledSym).flatMap {
+          case (eff, handlers) =>
+            handlers.zipWithIndex.flatMap {
+              case (handler1, i) => handlers.drop(i+1).map
+              {
+                case handler2 => EntryPointError.DuplicatedDefaultHandlers(eff, handler1.handlerDef.loc, handler2.handlerDef.loc)
+              }
+            }
+        }
+      (root, errors.toList.flatMap(_.toList) ++ duplicateHandlersErrs)
+    }
+
+    /**
+      * Validates that a function marked with @DefaultHandler has the correct signature and structure.
+      *
+      * A valid default handler must:
+      * - Be in the companion module of an existing effect
+      * - Take exactly one argument of function type: Unit -> a \ ef
+      * - Return the same type variable 'a' as the function argument
+      * - Have an effect signature of the form: (ef - HandledEffect) + PrimitiveEffect1 + ...
+      *
+      * Valid handler example
+      * {{{
+      * eff efHandle
+      * mod efHandle {
+      *   @DefaultHandler
+      *   def handle(f: Unit -> a \ ef): a \ (ef - efHandle) + PrimitiveEffect1 + ... = exp
+      * }
+      * }}}
+      *
+      * Invalid handler example
+      * {{{
+      * @DefaultHandler
+      * def handle(f: Unit -> a \ ef, arg: tpe): Bool \ (ef - ef2) + PrimitiveEffect1 + ... = exp
+      * }}}
+      *
+      * @param handlerSym The symbol of the handler function
+      * @param handlerDef The definition of the handler function
+      * @param root       The typed AST root
+      * @return Either a valid HandlerInfo or a list of entry point errors
+      */
+    private def checkHandler(handlerSym: Symbol.DefnSym, handlerDef: TypedAst.Def, root: TypedAst.Root)(implicit flix: Flix): Validation[HandlerInfo, EntryPointError] = {
+      // First we obtain the effect corresponding to the handler
+      val effFqn = handlerSym.namespace.mkString(".")
+      val effSymbol = Symbol.mkEffSym(effFqn)
+      val companionEffect = root.effects.get(effSymbol)
+      companionEffect match {
+        // There is a valid effect to wrap
+        case Some(_) =>
+          // Check that handler has a signature like
+          // def handle(f: Unit -> a \ ef): a \ (ef - efHandle) + PrimitiveEffect1 + ...
+          val efHandle = Type.Cst(TypeConstructor.Effect(effSymbol, Kind.Eff), SourceLocation.Unknown)
```

Comment:
Instead of using Unknown, the best would be to grab the loc of handlerSym and then use the synthethic version of that. Do that everywhere instead of Unknown.